### PR TITLE
feat: support postgresql

### PR DESCRIPTION
**Pull request by holicc**
#177

### CODE REVIEW

File: columnq/Cargo.toml [hunk 1]
```diff
@@ -68,31 +69,11 @@ rusqlite = "0"
...
-    "yup-oauth2/hyper-tls",
-]
-native-tls = [
-    "reqwest/native-tls",
-    "hyper-tls",
-    "deltalake/s3",
-    "deltalake/azure",
-    "yup-oauth2/hyper-tls",
-]
+rustls = ["hyper-rustls", "reqwest/rustls-tls", "deltalake/s3", "deltalake/azure", "yup-oauth2/hyper-rustls"]
```
Comment (houqp) [inline columnq/Cargo.toml:72]: nit, could we change it back to multi-line formatting to keep diffing easier to follow?

File: columnq/src/table/database.rs [hunk 2]
```diff
@@ -73,23 +93,63 @@ mod imp {
...
+                            .map_err(|e| ColumnQError::Database(e.to_string()))?;
+                        let tls = match config.get_ssl_mode() {
+                            tokio_postgres::config::SslMode::Require => tokio_postgres::NoTls,
+                            _ => tokio_postgres::NoTls,
+                        };
+                        let source: PostgresSource<BinaryProtocol, tokio_postgres::NoTls> =
+                            PostgresSource::new(config.into(), tls, 2)
+                                .map_err(|e| ColumnQError::Database(e.to_string()))?;
+                        let queries = queries.clone();
+                        let task = tokio::task::spawn_blocking(move || {
```
Comment (houqp) [inline columnq/src/table/database.rs:108]: I might be missing something here, why do we need to fetch postgres data with blocking code here?

File: columnq/Cargo.toml (hunk 1 above)
Comment (holicc) [reply columnq/Cargo.toml:72]: sure

File: columnq/src/table/database.rs (hunk 2 above)
Comment (holicc) [reply columnq/src/table/database.rs:108]: if we don't block there, will cause err: `Cannot start a runtime from within a runtime`.

Comment (houqp) [reply columnq/src/table/database.rs:108]: This is specific to the PostgresSource? I will take a closer look at it this weekend, thanks for the patch!

If this is expected, then we should also file an upstream bug report, async call context should be supported as first class citizen in connectorx.

Comment (holicc) [reply columnq/src/table/database.rs:108]: yes, it's specific to the `PostgresSource`.

**Comment by yuyang-ok** [comment]
@houqp Is this going to merge? I aslo have interesting about this

File: columnq/src/table/database.rs (hunk 2 above)
Comment (yuyang-ok) [reply columnq/src/table/database.rs:109]: @holicc Do you know why have this error `Cannot start a runtime from within a runtime`?

Comment (holicc) [reply columnq/src/table/database.rs:109]: In somewhere create a new runtime when using PgSource

**Comment by houqp** [comment]
sorry for the delay, will take a look at this later today

**Comment by houqp** [comment]
@holicc do you need help fixing the CI? I was planning to dig into that, but don't want to step on your toes.

**Comment by holicc** [comment]
> @holicc do you need help fixing the CI? I was planning to dig into that, but don't want to step on your toes.

Yes, I'm not sure how to fix it.

**Comment by houqp** [comment]
@holicc this is because the gcs setup in CI was failing, for some reasons, the curl command was not able to store the parquet file there. Looking into it more now.

**Comment by houqp** [comment]
hopefully https://github.com/roapi/roapi/pull/291 should fix it.

**Comment by houqp** [comment]
@holicc could you rebase to get a clean CI run?

**Comment by houqp** [comment]
OK, now we are back to MacOS openssl cross build issues :(

@holicc I think the quickest fix is to disable the postgres feature in the mac build job: https://github.com/roapi/roapi/blob/b8b772a01853b3014b47d6242be111ee2e2c7549/.github/workflows/build.yml#L292.

**Comment by holicc** [comment]
> OK, now we are back to MacOS openssl cross build issues :(
> 
> @holicc I think the quickest fix is to disable the postgres feature in the mac build job:
> 
> https://github.com/roapi/roapi/blob/b8b772a01853b3014b47d6242be111ee2e2c7549/.github/workflows/build.yml#L292
> 
> .

It appears that we need to install `openssl` in the build environment.

- https://github.com/sfackler/rust-openssl/issues/855
- https://ma.ttias.be/could-not-find-directory-of-openssl-installation/

**Comment by houqp** [comment]
@holicc I was suggesting we don't use the `--features database` in the macos crossbuild job, but instead build explicitly with the following flags to exclude postgres for mac for now:
* database-sqlite
* database-mysql

Cross building openssl on mac is going to be a major undertake unfortunately. So we can look into fixing that later.

**Comment by holicc** [comment]
> @holicc I was suggesting we don't use the `--features database` in the macos crossbuild job, but instead build explicitly with the following flags to exclude postgres for mac for now:
> 
> * database-sqlite
> * database-mysql
> 
> Cross building openssl on mac is going to be a major undertake unfortunately. So we can look into fixing that later.

My bad. I thought it was a simple problem.

**Review by houqp — APPROVED** [review]

**Comment by houqp** [comment]
Thank you @holicc !

**Comment by holicc** [comment]
Recently, I found that we can simplify the code base by using the get_arrow API from connectorx.
 @houqp Do you think we should modify the code like this?

```rust
impl DatabaseLoader {
        pub fn to_mem_table(
            &self,
            t: &TableSource,
        ) -> Result<datafusion::datasource::MemTable, ColumnQError> {
            debug!("loading database table data...");
            let queries = CXQuery::naked(format!("SELECT * FROM {}", t.name));
            let source = SourceConn::try_from(t.get_uri_str())
                .map_err(|e| ColumnQError::Database(e.to_string()))?;
            let destination = connectorx::get_arrow::get_arrow(&source, None, &[queries])
                .map_err(|e| ColumnQError::Database(e.to_string()))?;
            Ok(datafusion::datasource::MemTable::try_new(
                destination.arrow_schema(),
                vec![destination.arrow().unwrap()],
            )?)
        }
    }
```

**Comment by houqp** [comment]
Yes, let's do that @holicc !

**Pull request merged**